src/com/esotericsoftware/kryo/serializers/CollectionSerializer.java

```diff
@@ -170,12 +171,11 @@ protected void writeHeader (Kryo kryo, Output output, T collection) {
 
 	/** Used by {@link #read(Kryo, Input, Class)} to create the new object. This can be overridden to customize object creation (eg
 	 * to call a constructor with arguments), optionally reading bytes written in {@link #writeHeader(Kryo, Output, Collection)}.
-	 * The default implementation uses {@link Kryo#newInstance(Class)} with special cases for ArrayList. */
+	 * The default implementation uses {@link Kryo#newInstance(Class)} with special cases for ArrayList and HashSet. */
 	protected T create (Kryo kryo, Input input, Class<? extends T> type, int size) {
-		if (type == ArrayList.class) return (T)new ArrayList(size);
-		T collection = kryo.newInstance(type);
-		if (collection instanceof ArrayList) ((ArrayList)collection).ensureCapacity(size);

```

Comment:
I removed this condition because it is confusing We return early for `ArrayList` above and this condition can only ever match for subclasses of `ArrayList`. Do we really care about `ArrayList` sub-classes?

Reply:
I don't find it very confusing. Is there a reason to not care about subclasses?
So I don't have a strong opinion here, but would prefer to keep this.

Reply:
I have no strong opinion about this either. Let's keep it.

Reply:
Done